lib/db_connection.ex

```diff
@@ -738,45 +736,134 @@ defmodule DBConnection do
   @spec run(conn, (t -> result), opts :: Keyword.t) :: result when result: var
   def run(conn, fun, opts \\ [])
   def run(%DBConnection{} = conn, fun, _) do
-    case get_info(conn) do
-      {_, _} ->
+    case fetch_conn(conn, nil) do
+      {:ok, _, _} ->
         fun.(conn)
-      :closed ->
-        raise DBConnection.ConnectionError, "connection is closed"
+      {:error, err, _} ->
+        raise err
     end
   end
   def run(pool, fun, opts) do
+    conn = checkout!(pool, opts)
+    try do
+      fun.(conn)
+    after
+      checkin(conn, opts)
+    end
+  end
+  @doc """
+  Acquire a lock on a connection.
+
+  Returns `{:ok, conn}` on success or `{;error, exception}` if there is an
+  error.
+
+  To use the locked connection call requests with the connection
+  reference returned. If the connection disconnects all future calls using that
+  connection reference will fail.
```

Comment:
What happens if I call it multiple times from the same process? Error? I get multiple connections? Other?